cpp/src/arrow/buffer.h

```diff
@@ -254,20 +252,53 @@ class ARROW_EXPORT Buffer {
   /// `data()` pointer.  Otherwise, you'll have to `View()` or `Copy()` it.
   bool is_cpu() const { return is_cpu_; }
 
+  /// \brief Whether the buffer is mutable
+  ///
+  /// If this function returns true, you are allowed to modify buffer contents
+  /// using the pointer returned by `mutable_data()` or `mutable_address()`.
+  bool is_mutable() const { return is_mutable_; }
+
   const std::shared_ptr<Device>& device() const { return memory_manager_->device(); }
 
   const std::shared_ptr<MemoryManager>& memory_manager() const { return memory_manager_; }
 
   std::shared_ptr<Buffer> parent() const { return parent_; }
 
-  // Convenience functions
+  /// \brief Get a RandomAccessFile for reading a buffer
+  ///
+  /// The returned file object reads from this buffer's underlying memory.
   static Result<std::shared_ptr<io::RandomAccessFile>> GetReader(std::shared_ptr<Buffer>);
+
+  /// \brief Get a OutputStream for writing to a buffer
+  ///
+  /// The buffer must be mutable.  The returned stream object writes into the buffer's
+  /// underlying memory (but it won't resize it).
   static Result<std::shared_ptr<io::OutputStream>> GetWriter(std::shared_ptr<Buffer>);
 
+  /// \brief Copy buffer
+  ///
+  /// The buffer contents will be copied into a new buffer allocated by the
+  /// given MemoryManager.  This function supports cross-device copies.
   static Result<std::shared_ptr<Buffer>> Copy(std::shared_ptr<Buffer> source,
                                               const std::shared_ptr<MemoryManager>& to);
+
+  /// \brief View buffer
+  ///
+  /// Return a Buffer that reflects this buffer, seen potentially from another
+  /// device, without making an explicit copy of the contents.  The underlying
+  /// mechanism is typically implemented by the kernel or device driver, and may
+  /// involve lazy caching of parts of the buffer contents on the destination
+  /// device's memory.
+  ///
+  /// This function may fail, since not all buffers support being viewed on
```

Comment:
A user might reasonably expect that "fail" means "return an error status", like `NotImplemented("CPU buffers viewed on FPGA")`.

In fact: when a noncopy view is unsupported then null will be returned and an error is only expected when it's bubbled up from cuda or so.

Reply:
Yes, I'm going to improve this. Thank you.